crates/gitbutler-branch-actions/src/branch_upstream_integration.rs

```diff
@@ -15,6 +15,108 @@ use crate::{
     conflicts, VirtualBranchesExt as _,
 };
 
+pub fn integrate_upstream_commits_for_series(
+    ctx: &CommandContext,
+    branch_id: StackId,
+    perm: &mut WorktreeWritePermission,
+    series_name: String,
+) -> Result<()> {
+    conflicts::is_conflicting(ctx, None)?;
+
+    let repository = ctx.repository();
+    let vb_state = ctx.project().virtual_branches();
+
+    let branch = vb_state.get_branch_in_workspace(branch_id)?;
+    let all_series = branch.list_series(ctx)?;
+
+    let default_target = vb_state.get_default_target()?;
+    let remote = default_target.clone().push_remote_name.ok_or(anyhow!(
+        "No remote has been configured for the target branch"
+    ))?;
+
+    let series = all_series
+        .iter()
+        .find(|series| series.head.name == series_name)
+        .ok_or(anyhow!("Series not found"))?;
+    let upstream_reference = series.head.remote_reference(remote.as_str())?;
+    let remote_head = ctx
+        .repository()
+        .find_reference(&upstream_reference)?
+        .peel_to_commit()?;
+    let stack_merge_base = ctx
+        .repository()
+        .merge_base(branch.head(), default_target.sha)?;
+    let upstream_to_local_merge_base = ctx
+        .repository()
+        .merge_base(branch.head(), remote_head.id())?;
+
+    // commits to rebase
+    let mut ordered_commits_to_rebase = vec![];
+    for series in all_series.clone() {
+        // if this is the series that is getting the upstream commits, they are added first
+        if series.head.name == series_name {
+            if branch.allow_rebasing {
+                // If rebase is allowed, add the upstream only commits
+                for id in series.upstream_only(&all_series).iter() {
+                    let commit = commit_by_oid_or_change_id(
+                        id,
+                        ctx,
+                        remote_head.id(),
+                        upstream_to_local_merge_base,
+                    )?;
+                    ordered_commits_to_rebase.push(commit.id());
+                }
+            } else {
+                // If merge is preferred, create a merge commit between the remote and the series head and add that
+                let series_head_commit = commit_by_oid_or_change_id(
+                    &series.head.target,
+                    ctx,
+                    branch.head(),
+                    upstream_to_local_merge_base,
+                )?;
+                let merge_commit = gitbutler_merge_commits(
```

Comment:
now that i think about it - i shouldnt be doing this at all. i think i have an idea how to approach this.

In a nutshell: if this is a stack of one (or the top of the stack series), allow making the merge commit and update the head (the head of the series will equal the head of the stack in this case).

Otherwise refuse to do this and tell the user that if they are doing stacking rebasing is a must